File: esthesis-core-ui/src/app/app.constants.ts
import {Icon, icon} from "leaflet";

export const AppConstants = {
  // The API root context.
  API_ROOT: "/api",

  // An id designating a new record is to be created.
  NEW_RECORD_ID: "new",

  // The date format to use.
  DATE_FORMAT: "yyyy-MM-dd HH:mm:ss",

  // The different types of security keys and CERTIFICATE.
  KEY_TYPE: {
    PRIVATE_KEY: "PRIVATE",
    PUBLIC_KEY: "PUBLIC",
    CERTIFICATE: "CERTIFICATE"
  },

  // A list of application-wide settings (see also the counterpart Java class AppConstants.java).
  NAMED_SETTING: {
    DEVICE_PROVISIONING_URL: "DEVICE_PROVISIONING_URL",
    DEVICE_PROVISIONING_SECURE: "DEVICE_PROVISIONING_SECURE",
    DEVICE_PROVISIONING_SEMVER: "DEVICE_PROVISIONING_SEMVER",
    DEVICE_REGISTRATION_MODE: "DEVICE_REGISTRATION_MODE",
    DEVICE_REGISTRATION_SECRET: "DEVICE_REGISTRATION_SECRET",
    DEVICE_ROOT_CA: "DEVICE_ROOT_CA",
    KAFKA_TOPIC_COMMAND_REQUEST: "KAFKA_TOPIC_COMMAND_REQUEST",
    DEVICE_PROVISIONING_CACHE_TIME: "DEVICE_PROVISIONING_CACHE_TIME",
    DEVICE_PUSHED_TAGS: "DEVICE_PUSHED_TAGS"
  },

  // Keystore items.
  KEYSTORE: {
    ITEM: {
      RESOURCE_TYPE: {
        DEVICE: "DEVICE",
        CERTIFICATE: "CERT",
        CA: "CA",
        TAG: "TAG"
      },
      KEY_TYPE: {
        PRIVATE_KEY: "PRIVATE",
        CERTIFICATE: "CERT"
      }
    }
  },

  // Device settings.
  DEVICE: {
    ATTRIBUTE: {
      TYPE: {
        STRING: "STRING",
        BOOLEAN: "BOOLEAN",
        BYTE: "BYTE",
        SHORT: "SHORT",
        INTEGER: "INTEGER",
        LONG: "LONG",
        BIG_DECIMAL: "BIG_DECIMAL",
        UNKNOWN: "UNKNOWN"
      }
    },
    TYPE: {
      ESTHESIS: "ESTHESIS"
    },
    STATUS: {
      DISABLED: "DISABLED",
      PREREGISTERED: "PREREGISTERED",
      REGISTERED: "REGISTERED",
      APPROVAL: "APPROVAL"
    },
    SETTING: {
      DEVICE_GEO_LAT: "DEVICE_GEO_LAT",
      DEVICE_GEO_LON: "DEVICE_GEO_LON"
    },
    COMMAND: {
      TYPE: {
        EXECUTE: "e",
        HEALTH: "h",
        PING: "p",
        FIRMWARE: "f",
        REBOOT: "r",
        SHUTDOWN: "s"
      },
      EXECUTION: {
        SYNCHRONOUS: "s",
        ASYNCHRONOUS: "a"
      }
    },
    DATA_IMPORT: {
      TYPE: {
        TELEMETRY: "TELEMETRY",
        METADATA: "METADATA"
      },
      SOURCE: {
        TEXT: "TEXT",
        FILE: "FILE"
      }
    }
  },

  // User statuses.
  USER_STATUS: {
    DISABLED: 0,
    ENABLED: 1
  },

  // Provisioning.
  PROVISIONING: {
    TYPE: {
      INTERNAL: "INTERNAL",
      EXTERNAL: "EXTERNAL"
    }
  },

  // Campaign constants.
  CAMPAIGN: {
    STATE: {
      CREATED: "CREATED",
      RUNNING: "RUNNING",
      PAUSED_BY_USER: "PAUSED_BY_USER",
      PAUSED_BY_WORKFLOW: "PAUSED_BY_WORKFLOW",
      TERMINATED_BY_WORKFLOW: "TERMINATED_BY_WORKFLOW",
      TERMINATED_BY_USER: "TERMINATED_BY_USER",
    },
    TYPE: {
      PROVISIONING: "PROVISIONING",
      EXECUTE_COMMAND: "EXECUTE_COMMAND",
      REBOOT: "REBOOT",
      SHUTDOWN: "SHUTDOWN",
      PING: "PING",
    },
    CONDITION: {
      TYPE: {
        DATETIME: "DATETIME",
        SUCCESS: "SUCCESS",
        PROPERTY: "PROPERTY",
        PAUSE: "PAUSE",
        BATCH: "BATCH",
      },
      STAGE: {
        ENTRY: "ENTRY",
        EXIT: "EXIT",
        INSIDE: "INSIDE"
      },
      OP: {
        BEFORE: "BEFORE",
        AFTER: "AFTER",
        FOREVER: "FOREVER",
        TIMER_MINUTES: "TIMER_MINUTES",
        EQUAL: "EQUAL",
        GT: "GT",
        LT: "LT",
        GTE: "GTE",
        LTE: "LTE",
      }
    },
    MEMBER_TYPE: {
      DEVICE: "DEVICE",
      TAG: "TAG",
    }
  },

  DIALOG_RESULT: {
    CANCEL: 0,
    SAVE: 1,
    DELETE: 2,
    OK: 3
  },

  SECURITY: {
    // The name of session keys where user data is saved.
    SESSION_STORAGE: {
      USERDATA: "X-ESTHESIS-USERDATA",
      PERMISSIONS: "X-ESTHESIS-PERMISSIONS"
    },
    // The name of fields in user data.
    USERDATA: {
      USERNAME: "preferred_username",
      FIRST_NAME: "given_name",
      LAST_NAME: "family_name"
    },
    // ERN
    ERN: {
      ROOT: "ern",
      SYSTEM: "esthesis",
      SUBSYSTEM: "core"
    },
    // Equivalent to AppConstants.Security.Category
    CATEGORY: {
      ABOUT: "about",
      APPLICATION: "application",
      AUDIT: "audit",
      CA: "ca",
      CAMPAIGN: "campaign",
      CERTIFICATES: "certificates",
      COMMAND: "command",
      CRYPTO: "crypto",
      DASHBOARD: "dashboard",
      DATAFLOW: "dataflow",
      DEVICE: "device",
      INFRASTRUCTURE: "infrastructure",
      KEYSTORE: "keystore",
      KUBERNETES: "kubernetes",
      PROVISIONING: "provisioning",
      SECURITY: "security",
      SETTINGS: "settings",
      TAGS: "tags",
      USERS: "users",
      GROUPS: "groups",
      ROLES: "roles",
      POLICIES: "policies"
    },
    // Equivalent to AppConstants.Security.Operation
    OPERATION: {
      CREATE: "create",
      READ: "read",
      WRITE: "write",
      DELETE: "delete"
    },
    PERMISSION: {
      ALLOW: "allow",
      DENY: "deny"
    }
  },

  DASHBOARD: {
    REFRESH_INTERVAL_MINUTES: 1, // How often to send a keep-alive.
    DEFAULTS: {
      COLUMN_WIDTH: 100
    },
    SINGLE_INSTANCE_ITEMS: [""],
    ITEM: {
      TYPE: {
        ABOUT: "ABOUT",
        AUDIT: "AUDIT",
        CAMPAIGNS: "CAMPAIGNS",
        DATETIME: "DATETIME",
        DEVICE_MAP: "DEVICE_MAP",
        DEVICES_LAST_SEEN: "DEVICES_LAST_SEEN",
        DEVICES_LATEST: "DEVICES_LATEST",
        DEVICES_STATUS: "DEVICES_STATUS",
        IMAGE: "IMAGE",
        NOTES: "NOTES",
        SECURITY_STATS: "SECURITY_STATS",
        SENSOR: "SENSOR",
        SENSOR_ICON: "SENSOR_ICON",
        TITLE: "TITLE"
      },
      COLUMNS: {
        ABOUT: 4,
        AUDIT: 4,
        CAMPAIGNS: 5,
        DATETIME: 4,
        DEVICE_MAP: 4,
        DEVICES_LAST_SEEN: 3,
        DEVICES_LATEST: 5,
        DEVICES_STATUS: 8,
        IMAGE: 5,
        NOTES: 4,
        SECURITY_STATS: 6,
        SENSOR: 2,
        SENSOR_ICON: 2,
        TITLE: 4
      },
      DEFAULTS: {
        ABOUT: {},
        AUDIT: {
          entries: 5
        },
        CAMPAIGNS: {},
        DATETIME: {
          date: true,
          time: true,
          local: true,
          server: true,
          formatDate: "MMMM Do YYYY",
          formatTime: "h:mm:ss a",
          formatDateTime: "MMMM Do YYYY, h:mm:ss a"
        },
        DEVICE_MAP: {
          zoom: 2,
          height: 200,
          hardwareIds: [],
          tags: []
        },
        DEVICES_LAST_SEEN: {},
        DEVICES_LATEST: {
          entries: 5
        },
        DEVICES_STATUS: {
          orientation: "horizontal"
        },
        IMAGE: {
          height: 400,
          refresh: 0,
          imageUrl: "https://placehold.co/450x400?text=esthesis%20IoT"
        },
        NOTES: {
          notes: "esthesis IoT"
        },
        SECURITY_STATS: {
          orientation: "horizontal"
        },
        SENSOR: {},
        SENSOR_ICON: {},
        TITLE: {
          title: "esthesis IoT"
        }
      }
    }
  },

  MAP_DEFAULT_ICON: icon({
    ...Icon.Default.prototype.options,
    iconUrl: "assets/marker-icon.png",
    iconRetinaUrl: "assets/marker-icon-2x.png",
    shadowUrl: "assets/marker-shadow.png"
  }),

  // Get a value by key name.
  getValue<T = any>(keyPath: string): T | undefined {
    return keyPath
    .split('.')
    .reduce((obj: any, key) => obj?.[key], AppConstants);
  }

};

File: esthesis-core-backend/common/src/main/java/esthesis/core/common/AppConstants.java
package esthesis.core.common;

/**
 * Global application constants across all CORE modules.
 */
public class AppConstants {

	// A keyword used to indicate a new record being created.
	public static final String NEW_RECORD_ID = "new";

	// Redis key suffixes for measurements.
	public static final String REDIS_KEY_SUFFIX_VALUE_TYPE = "valueType";
	public static final String REDIS_KEY_SUFFIX_TIMESTAMP = "timestamp";
	public static final String REDIS_KEY_PROVISIONING_PACKAGE_FILE = "file";

	// Naming convention for hardware ids.
	public static final String HARDWARE_ID_REGEX = "^[a-zA-Z0-9_-]*$";

	// The size limit when displaying possibly large content in the logs.
	public static final int MESSAGE_LOG_ABBREVIATION_LENGTH = 4096;

	// Security roles.
	public static final String ROLE_USER = "user";
	public static final String ROLE_SYSTEM = "system";

	// Settings keys.
	public enum NamedSetting {
		SECURITY_ASYMMETRIC_KEY_SIZE, SECURITY_ASYMMETRIC_KEY_ALGORITHM,
		SECURITY_ASYMMETRIC_SIGNATURE_ALGORITHM,

		DEVICE_ROOT_CA, DEVICE_PROVISIONING_URL, DEVICE_REGISTRATION_MODE,
		DEVICE_REGISTRATION_SECRET, DEVICE_PROVISIONING_SEMVER,
		DEVICE_TAGS_ALGORITHM, DEVICE_GEO_LAT, DEVICE_GEO_LON,
		DEVICE_PROVISIONING_SECURE, DEVICE_PROVISIONING_CACHE_TIME,
		DEVICE_PUSHED_TAGS,

		KAFKA_TOPIC_COMMAND_REQUEST, KAFKA_TOPIC_TELEMETRY, KAFKA_TOPIC_METADATA,
		KAFKA_TOPIC_APPLICATION,

		CHATBOT_ENABLED
	}

	// Types of security keys.
	public enum KeyType {
		PUBLIC, PRIVATE, CERTIFICATE
	}

	// The registration mode available in CORE.
	public enum DeviceRegistrationMode {
		DISABLED, OPEN, OPEN_WITH_SECRET, ID
	}

	// Device-related constants.
	public static class Device {

		public enum Status {
			DISABLED, PREREGISTERED, REGISTERED
		}

		public enum DataImportType {
			TELEMETRY, METADATA
		}

		public static class Attribute {

			public enum Type {
				STRING, BOOLEAN, BYTE, SHORT, INTEGER, LONG, BIG_DECIMAL, UNKNOWN
			}
		}

	}

	// MongoDB GridFS constants.
	public static class GridFS {

		private GridFS() {
		}

		public static final String PROVISIONING_BUCKET_NAME = "ProvisioningPackageBucket";
		public static final String PROVISIONING_METADATA_NAME = "provisioningPackageId";
	}

	public static class Keystore {

		private Keystore() {
		}

		public static class Item {

			public enum ResourceType {
				DEVICE, CERT, CA, TAG
			}

			public enum KeyType {
				PRIVATE, CERT
			}
		}
	}

	// Constants to identify provisioning packages and cache them in Redis.
	public static class Provisioning {

		public enum Type {
			INTERNAL, EXTERNAL
		}

		public static class Redis {

			public static final String DOWNLOAD_TOKEN_PACKAGE_ID = "DTPI";
			public static final String DOWNLOAD_TOKEN_CREATED_ON = "DTCO";

			private Redis() {
			}
		}
	}

	// Security and Audit constants.
	public static class Security {

		public enum Category {
			ABOUT,
			APPLICATION,
			AUDIT,
			CA,
			CAMPAIGN,
			CERTIFICATES,
			COMMAND,
			CRYPTO,
			DASHBOARD,
			DATAFLOW,
			DEVICE,
			INFRASTRUCTURE,
			KEYSTORE,
			KUBERNETES,
			PROVISIONING,
			SECURITY,
			SETTINGS,
			TAGS,
			USERS,
			GROUPS,
			ROLES,
			POLICIES,
			NULL  // Required for annotations.
		}

		// Available operations.
		public enum Operation {
			CREATE, READ, WRITE, DELETE, AUDIT,
			NULL  // Required for annotations.
		}

		// Available permissions.
		public enum Permission {

			ALLOW, DENY;
		}

		// ERN common prefixes.
		public static class Ern {

			private Ern() {
			}

			public static final String ROOT = "ern";
			public static final String SYSTEM = "esthesis";
			public static final String SUBSYSTEM = "core";
		}
	}

	// Constants for the Campaigns functionality.
	public static class Campaign {

		public enum State {

			CREATED, RUNNING, PAUSED_BY_USER, PAUSED_BY_WORKFLOW,
			TERMINATED_BY_WORKFLOW, TERMINATED_BY_USER;
		}

		public enum Type {

			PROVISIONING, EXECUTE_COMMAND, REBOOT, SHUTDOWN, PING;
		}

		public static class Member {

			public enum Type {
				DEVICE, TAG;
			}
		}

		public static class Condition {

			public enum Type {
				DATETIME, SUCCESS, PROPERTY, PAUSE, BATCH;
			}

			public enum Stage {
				ENTRY, EXIT, INSIDE;
			}

			public enum Op {

				BEFORE, AFTER, ABOVE, BELOW, FOREVER, TIMER_MINUTES, EQUAL,
				GT, LT, GTE, LTE;
			}
		}
	}

	// Constants for the Dashboard functionality.
	public static class Dashboard {

		public enum Type {
			ABOUT, AUDIT, CAMPAIGNS, CHART, DATETIME, DEVICE_MAP, DEVICES_LAST_SEEN, DEVICES_LATEST,
			DEVICES_STATUS, DIFF, IMAGE, NOTES, SECURITY_STATS, SENSOR, SENSOR_ICON, TITLE
		}
	}
}
